src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/trait/BitbucketPullRequestDiscoveryTrait.java

```diff
@@ -0,0 +1,166 @@
+import com.atlassian.bitbucket.jenkins.internal.client.BitbucketRepositoryClient;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketPluginConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketServerConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.credentials.JenkinsToBitbucketCredentials;
+import com.atlassian.bitbucket.jenkins.internal.model.BitbucketPullRequestState;
+import com.atlassian.bitbucket.jenkins.internal.scm.*;
+import hudson.Extension;
+import hudson.plugins.git.UserMergeOptions;
+import hudson.plugins.git.extensions.impl.PreBuildMerge;
+import jenkins.plugins.git.GitSCMBuilder;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.mixin.ChangeRequestCheckoutStrategy;
+import jenkins.scm.api.trait.SCMBuilder;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceTrait;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import jenkins.scm.impl.trait.Discovery;
+import org.jenkinsci.plugins.gitclient.MergeCommand;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.inject.Inject;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+/**
+ * @since 4.0.0
+ */
+public class BitbucketPullRequestDiscoveryTrait extends SCMSourceTrait {
+
+    private static final Logger log = Logger.getLogger(BitbucketPullRequestDiscoveryTrait.class.getName());
+
+    @DataBoundConstructor
+    public BitbucketPullRequestDiscoveryTrait() {
+    }
+
+    @Override
+    protected void decorateBuilder(SCMBuilder<?, ?> builder) {
+        if (builder instanceof GitSCMBuilder) {
+            GitSCMBuilder<?> gitSCMBuilder = (GitSCMBuilder<?>) builder;
+            SCMRevision revision = gitSCMBuilder.revision();
+
+            if (revision instanceof BitbucketPullRequestSCMRevision) {
+                BitbucketPullRequestSCMRevision prRevision = (BitbucketPullRequestSCMRevision) revision;
+                BitbucketPullRequestSCMHead prHead = (BitbucketPullRequestSCMHead) prRevision.getHead();
+
+                // The BitbucketPullRequestSCMHead uses the PR id as the head name, so we need to use a custom
+                // refspec to be able to map to the correct PR refs during checkout.
+                gitSCMBuilder.withRefSpec("+refs/pull-requests/" + prHead.getId() + "/from" +
+                        ":refs/remotes/@{remote}/" + prHead.getName());
+
+                // Additionally, we also need to add the source branch name the underlying GitSCM's environment
+                // so it can be easily referenced if needed.
+                gitSCMBuilder.withExtension(new BitbucketPullRequestSourceBranch(prHead.getPullRequest()));
+
+                // Use the merge checkout strategy if the head specifies it
+                if (prHead.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) {
```

Comment:
again a log if this check fails seems like a good thing from a supportability point of view